experiments/tf_trainer/tf_kona_prototypical_network/proto.py

```diff
@@ -0,0 +1,317 @@
+  iterator = tf.data.Iterator.from_string_handle(handle, output_types,
+                                                 output_shapes)
+  episode_batch = iterator.get_next()
+
+  with tf.variable_scope("negative_prototype"):
+    negative_prototype = get_prototype(episode_batch["negative_supports"])
+  with tf.variable_scope("positive_prototype"):
+    positive_prototype = get_prototype(episode_batch["positive_supports"])
+  with tf.variable_scope("negative_embeddings"):
+    negative_embeddings = get_embeddings(episode_batch["negative_queries"])
+  with tf.variable_scope("positive_embeddings"):
+    positive_embeddings = get_embeddings(episode_batch["positive_queries"])
+
+  negative_logits = calculate_logits(negative_embeddings, positive_prototype,
+                                     negative_prototype)
+  positive_logits = calculate_logits(positive_embeddings, positive_prototype,
+                                     negative_prototype)
+
+  return handle, negative_logits, positive_logits
+
+
+def train_operation(negative_logits, positive_logits):
+  negative_loss = tf.losses.softmax_cross_entropy(
+      tf.broadcast_to(tf.one_hot(0, 2), tf.shape(negative_logits)),
+      negative_logits)
+  positive_loss = tf.losses.softmax_cross_entropy(
+      tf.broadcast_to(tf.one_hot(1, 2), tf.shape(positive_logits)),
+      positive_logits)
+  loss = negative_loss + positive_loss
+
+  optimizer = tf.train.AdamOptimizer(0.001)
+  train = optimizer.minimize(loss)
+  return (train, loss)
+
+
+def predictions_and_metrics(negative_logits, positive_logits):
+  predict = lambda logits: tf.argmax(logits, axis=1)
+
+  negative_predictions = predict(negative_logits)
+  negative_labels = tf.fill(tf.shape(negative_predictions), 0)
+  positive_predictions = predict(positive_logits)
+  positive_labels = tf.fill(tf.shape(positive_predictions), 1)
+
+  probability = tf.nn.softmax(
+      tf.concat([negative_logits, positive_logits], -2), axis=-1)
+  labels = tf.concat([negative_labels, positive_labels], -1)
+  predictions = tf.concat([negative_predictions, positive_predictions], -1)
+
+  acc_op, update_acc_op = tf.metrics.accuracy(labels, predictions)
+  auc_op, update_auc_op = tf.metrics.auc(labels,
+                                         tf.gather(probability, 1, axis=-1))
+  return (predictions, acc_op, auc_op, update_acc_op, update_auc_op)
+
+
+if FLAGS.model_dir:
+  model_dir = FLAGS.model_dir
+else:
+  st = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
+  model_dir = "gs://conversationai-models/jjtan/transfer_learning/model/" + st
+print("Model dir: " + model_dir)
+save_path = model_dir + "/save/model.ckpt"
+metadata_path = model_dir + "/meta.txt"
+
+with tf.gfile.Open(metadata_path, "w") as f:
+  f.write("Encoding Layers: " + FLAGS.encoding_layers + "\n")
+
+output_types = {
+    "negative_queries": tf.string,
+    "negative_supports": tf.string,
+    "positive_queries": tf.string,
+    "positive_supports": tf.string
+}
+output_shapes = {
```

Comment:
Just noting these hardcoded values. This seems ok for now, but maybe put a todo to make these into a flag in the future?